src/Akka.Persistence.Redis/Query/Stages/AllEventsSource.cs

```diff
@@ -0,0 +1,348 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Akka.Actor;
+using Akka.Configuration;
+using Akka.Pattern;
+using Akka.Persistence.Query;
+using Akka.Streams;
+using Akka.Streams.Stage;
+using StackExchange.Redis;
+
+namespace Akka.Persistence.Redis.Query.Stages
+{
+    internal class AllEventsSource : GraphStage<SourceShape<EventEnvelope>>
+    {
+        private readonly ConnectionMultiplexer _redis;
+        private readonly int _database;
+        private readonly Config _config;
+        private readonly long _fromSequenceNr;
+        private readonly long _toSequenceNr;
+        private readonly ActorSystem _system;
+        private readonly bool _live;
+
+        public AllEventsSource(ConnectionMultiplexer redis, int database, Config config, long fromSequenceNr, long toSequenceNr, ActorSystem system, bool live)
+        {
+            _redis = redis;
+            _database = database;
+            _config = config;
+            _fromSequenceNr = fromSequenceNr;
+            _toSequenceNr = toSequenceNr;
+            _system = system;
+            _live = live;
+
+            Outlet = live
+                ? new Outlet<EventEnvelope>("AllEventsSource")
+                : new Outlet<EventEnvelope>("AllCurrentEventsSource");
+
+            Shape = new SourceShape<EventEnvelope>(Outlet);
+        }
+
+        internal Outlet<EventEnvelope> Outlet { get; }
+
+        public override SourceShape<EventEnvelope> Shape { get; }
+
+        protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes)
+        {
+            return new AllEventsLogic(_redis, _database, _config, _system, _fromSequenceNr, _toSequenceNr, _live, Outlet, Shape);
+        }
+
+        private enum State
+        {
+            Idle = 0,
```

Comment:
Can you describe how this state machine supposed to work? ie. what is its starting state, conditions of state change, what each state supposed to do?

Reply:
`Idle` is the start state - I think it is a way to prevent running concurrent database queries. For a live query with refresh interval, the state MUST be `Idle`  to run a new query

Reply:
The `Querying` State is an indicator to `CompleteStage()` for non live query. While for Live query that causes the state to be changed to `WaitingForNotification`. At the state, the state is changed to `Idle` which either sends a new query to the database if the buffer is empty or delivers the queued messages.